fl4health/clients/fenda_client.py

```diff
@@ -1,12 +1,226 @@
+
+    def predict(self, input: torch.Tensor) -> Dict[str, torch.Tensor]:
+
+        preds = self.model(input)
+
+        if self.contrastive_loss_weight or self.perFCL_loss_weights:
+            preds["old_local_features"] = self.old_local_module.forward(input).view(len(preds["local_features"]), -1)
+            if self.perFCL_loss_weights:
+                preds["old_global_features"] = self.old_global_module.forward(input).reshape(
+                    len(preds["global_features"]), -1
+                )
+                preds["aggregated_global_features"] = self.aggregated_global_module.forward(input).reshape(
+                    len(preds["global_features"]), -1
+                )
+        return preds
+
+    def set_parameters(self, parameters: NDArrays, config: Config) -> None:
+
+        # Save the parameters of the old model
+        assert isinstance(self.model.local_module, torch.nn.Module)
+        assert isinstance(self.model.global_module, torch.nn.Module)
+        if self.contrastive_loss_weight or self.perFCL_loss_weights:
+            self.old_local_module = self.clone_model(self.model.local_module)
+            self.old_global_module = self.clone_model(self.model.global_module)
+
+        # Set the parameters of the model
+        super().set_parameters(parameters, config)
+
+        # Save the parameters of the global model
+        if self.perFCL_loss_weights:
+            assert isinstance(self.model.global_module, torch.nn.Module)
+            self.aggregated_global_module = self.clone_model(self.model.global_module)
+
+        return
+
+    def get_cosine_similarity_loss(self, local_features: torch.Tensor, global_features: torch.Tensor) -> torch.Tensor:
+        """
+        Cosine similarity loss aims to minimize the similarity among current local features and current global
+        features of fenda model.
+        """
+        assert len(local_features) == len(global_features)
+        return torch.abs(self.cos_sim(local_features, global_features)).mean()
+
+    def get_contrastive_loss(
+        self, local_features: torch.Tensor, old_local_features: torch.Tensor, global_features: torch.Tensor
+    ) -> torch.Tensor:
+        """
+        Contrastive loss airms to enhance the similarity between the current local features and old local feature
+        as positive pairs while reducing the similarity between the current local features and current global
+        features as negative pairs.
+        """
+        assert isinstance(self.temperature, float)
+        assert len(local_features) == len(old_local_features)
+        posi = self.cos_sim(local_features, old_local_features)
+        logits = posi.reshape(-1, 1)
+        assert len(local_features) == len(global_features)
+        nega = self.cos_sim(local_features, global_features)
+        logits = torch.cat((logits, nega.reshape(-1, 1)), dim=1)
+        logits /= self.temperature
+        labels = torch.zeros(local_features.size(0)).to(self.device).long()
```

Comment:
Based on a bit of decomposition, I think the calculation you're doing here is correct, but it is a bit hard to read based on the manipulations and we perform pretty similar calculations in the get_perFCL_loss() below. So I think there is some merit to constructing a generic function that computes the `Q(., .)` between two vectors and the corresponding loss then calling them in these functions. Something like 
```
def contrastive_measure(z_1: torch.Tensor, z_2: torch.Tensor, temperature: float) -> torch.Tensor:
    similarity = self.cos_sim(z_1, z_2)
    scaled_sim = similarity/temperature
    return torch.exp(scaled_sim)

def constrastive_loss(reference: torch.Tensor, positive: torch.Tensor, negative: torch.Tensor, temperature: float) -> torch.Tensor:
    Q_num = contrastive_measure(reference, positive, temperature)
    Q_denom = contrastive_measure(reference, negative, temperature)

    loss = -torch.log(Q_num/(Q_num + Q_denom)).mean()
```
Then these can be called in various ways to form the losses?

Reply:
I would also recommend adding tests for each of the loss calculations to ensure they are being performed correctly.

Reply:
I get this kind of contrastive loss computation from MOON code, and I am not sure about changing it to these losses, as I have always used this version of supervised contrastive loss computing. Also this version has an extra torch.exp and torch.log that with this code we escape extra overhead of them. 
Although I can make a separate function to calculate contrastive loss and call it in PerFCL_loss function though, rather than compute it right there if it is so confusing now.

Reply:
I think it would be good to encapsulate it into a callable function that can be used by both this method and `get_perFCL_loss`. I'm good if we leave the calculation as is. It's just takes a bit of thinking to understand how the cross entropy loss is helping you here. There are likely optimizations in the torch loss, but in that loss, you still need to do the exponentiation and log calculations, its just being done by the torch library rather than by us.

Let's leave it as is (other than the encapsulation) and just make sure we write a few tests to verify the calculations are correct.